### ISSUE TITLE
Language specific folding

### ISSUE THREAD

**Issue by Kurren123** [Krzysztof-Cieslak/vscode-elm#278]
Long files in elm are considered pretty normal. Evan and others recommend organising long files with comments like `-- UPDATE`. 

Is it possible to have language specific folding so that I can make a comment to specify a region in a long file? Would make organisation much easier. 

I'm willing to lend a hand if this feature isn't yet implemented.

**Comment by Krzysztof-Cieslak** [comment]
Yes. it is possible to implement language specific folding in VSCode. To do it you'd need to implement [FoldingRangeProvider](https://code.visualstudio.com/docs/extensionAPI/vscode-api#FoldingRangeProvider). I'd be happy to merge PR adding this functionality :-)

**Comment by Kurren123** [comment]
So [looking at the docs](https://code.visualstudio.com/docs/extensionAPI/vscode-api#FoldingRangeKind) you can specify which folding ranges are comments, imports and regions. Here's what I'm thinking for now:

1. For comments we'll do multiline comments. Maybe later on we can look at grouping consecutive single line comments into one foldable comment block.
2. For imports we'll start at the first import in the file and end at the last. When using `elm-format` this should be good enough.
3. For regions there are two alternatives. 

**Option A**: we'll have two settings with the following default values:
```
{
  "startRegionRegex" : "^--- (.+)$",
  "endRegionRegex" : "^---\\s*$"
}
```
We'll use these to look for lines that should start and and a region. These defaults will allow you to do something like 

```
--- UPDATE

(code here)
...
...

---
```

Pros/cons: Allows nested regions, doesn't work with the convention in elm which goes something like  `-- UPDATE` or `-- VIEW`

**Option B**: We just have a `startRegionRegex : "^-- ([A-Z\\s]+)$" setting. Regions start when a line matches this regex and end either when another one starts or when the file ends.

Pros/cons: The default works immediately with most elm examples, where the region is a single line comment in all caps eg `-- MODEL`. However doesn't allow nested regions.

My vote is to keep things simple and go with option B. If lots of people want nested regions then we'll revisit from there.
